File: cpp/src/fen.cpp
#include "chronos/fen.h"
#include <sstream>
#include <stdexcept>
#include <cctype>

namespace chronos {

static void init_empty(Fen& f) {
    for (int r=0;r<8;r++) for (int c=0;c<8;c++) f.board[r][c]='.';
}

Fen Fen::parse(const std::string& fen_str) {
    Fen f;
    init_empty(f);

    std::istringstream iss(fen_str);
    std::string placement, stm, castling, ep;
    if (!(iss >> placement >> stm >> castling >> ep)) {
        throw std::runtime_error("Invalid FEN: missing fields");
    }
    iss >> f.halfmove_clock >> f.fullmove_number;

    // placement
    int r = 0, c = 0;
    for (char ch : placement) {
        if (ch == '/') {
            r++; c = 0;
            continue;
        }
        if (std::isdigit(static_cast<unsigned char>(ch))) {
            int n = ch - '0';
            for (int i=0;i<n;i++) {
                if (c>=8) throw std::runtime_error("Invalid FEN: file overflow");
                f.board[r][c++]='.';
            }
        } else {
            if (c>=8) throw std::runtime_error("Invalid FEN: file overflow");
            f.board[r][c++]=ch;
        }
    }
    if (r != 7) {
        // permissive but sanity check
    }

    // side to move
    if (stm == "w") f.white_to_move = true;
    else if (stm == "b") f.white_to_move = false;
    else throw std::runtime_error("Invalid FEN: side to move");

    // castling
    f.castle_wk = castling.find('K') != std::string::npos;
    f.castle_wq = castling.find('Q') != std::string::npos;
    f.castle_bk = castling.find('k') != std::string::npos;
    f.castle_bq = castling.find('q') != std::string::npos;

    return f;
}

} // namespace chronos

File: python/common/fen.py
from __future__ import annotations
from dataclasses import dataclass
from typing import Tuple

@dataclass
class Fen:
    # board ranks 8->1 as rows 0..7, files a->h as cols 0..7
    board: list[list[str]]
    white_to_move: bool
    castle_wk: bool
    castle_wq: bool
    castle_bk: bool
    castle_bq: bool
    halfmove_clock: int
    fullmove_number: int

PIECE_PLANES = {
    "P": 0, "N": 1, "B": 2, "R": 3, "Q": 4, "K": 5,
    "p": 6, "n": 7, "b": 8, "r": 9, "q": 10, "k": 11,
}

def parse_fen(fen_str: str) -> Fen:
    parts = fen_str.strip().split()
    if len(parts) < 4:
        raise ValueError("Invalid FEN: missing fields")

    placement, stm, castling, ep = parts[:4]
    halfmove = int(parts[4]) if len(parts) > 4 else 0
    fullmove = int(parts[5]) if len(parts) > 5 else 1

    board = [["." for _ in range(8)] for _ in range(8)]
    r = 0
    c = 0
    for ch in placement:
        if ch == "/":
            r += 1
            c = 0
            continue
        if ch.isdigit():
            n = int(ch)
            for _ in range(n):
                board[r][c] = "."
                c += 1
        else:
            board[r][c] = ch
            c += 1

    white_to_move = (stm == "w")
    castle_wk = "K" in castling
    castle_wq = "Q" in castling
    castle_bk = "k" in castling
    castle_bq = "q" in castling

    return Fen(
        board=board,
        white_to_move=white_to_move,
        castle_wk=castle_wk,
        castle_wq=castle_wq,
        castle_bk=castle_bk,
        castle_bq=castle_bq,
        halfmove_clock=halfmove,
        fullmove_number=fullmove,
    )
